### ISSUE TITLE
Question: FeadAuth cookie update / Access token renewal

### ISSUE THREAD

**Issue by sergey-tihon** [Yvand/AzureCP#62]
By default Azure AD generate Access token with 1 hour Lifetime.
https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-configurable-token-lifetimes#configurable-token-lifetime-properties

This means that `FedAuth` cookie will be valid only for 1h.
(If you do not have `Azure AD Premium` for [Conditional Access](https://docs.microsoft.com/en-us/azure/active-directory/conditional-access/howto-conditional-access-session-lifetime#configuring-authentication-session-controls) )

Technically, SharePoint reply 401 on the first call after Access Token expiration.

In the same `Refresh Token Max Inactive Time` is 90 days. If I understand it correctly, this mean that new access token can be acquired using the old one during this time frame.

# The question

Is it possible to silently request token renewal from Azure AD (when token is close to expire, inside LogonTokenCacheExpirationWindow) and set new FedAuth cookie header in the response to make user's life happier?

**Comment by sergey-tihon** [comment]
I am asking this because by default `AccessTokenLifetime`  is 1 hour. So, token expires every hour and SharePoint redirect us to AAD for login (not very pleasant user experience)

[Configurable token lifetimes in Azure Active Directory (Preview)](https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-configurable-token-lifetimes)  works till Nov 1, and will be replaced by [authentication session management with Conditional Access](https://docs.microsoft.com/en-us/azure/active-directory/conditional-access/howto-conditional-access-session-lifetime) that is premium feature that require additional licenses.

That is why usefulness of AzureCP for no-premium Azure AD users is questionable for me after Nov 1, 2019.

**Comment by Yvand** [comment]
This is interesting, I was not aware about this.
I agree that the default lifetime of 1h will be a painful limitation on non-premium Azure AD tenants after configurable token lifetime feature will be retired on November 1.
But I don't see what can be done on AzureCP/SharePoint side to deal with this. Silently refreshing the access token on behalf of the user is not possible, only the user can refresh it.

**Comment by sergey-tihon** [comment]
I see the update on MSDN: https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-configurable-token-lifetimes

> After hearing from customers during the preview, we've implemented authentication session management capabilities in Azure AD Conditional Access. You can use this new feature to configure refresh token lifetimes by setting sign in frequency. After **May 1, 2020** you will not be able to use Configurable Token Lifetime policy to configure session and refresh tokens. **You can still configure access token lifetimes after the deprecation**.

Does last sentence mean that access token lifetime will be configurable  after `May 1, 2020` and it is enough for AzureCP?

**Comment by Yvand** [comment]
@sergey-tihon can you please explain what is your concern?
When the access token expires, AzureCP (actually, Microsoft.Graph) has the plumbing to use the refresh token to acquire a new access token.
In my tests the access tokens are valid 1h, but I don't know about the refresh token.

**Comment by sergey-tihon** [comment]
it is not a concern, it is `new hope` 

I see the solution for my initial concern in following sentence
> You can still configure access token lifetimes after the deprecation.

Even after deprecation of `Configurable token lifetimes in Azure Active Directory (Preview)`, we will be able to increate `access token lifetimes` (even without Azure AD Premium).

This means that we can increase `access token lifetimes` from default `1h` to lets say `12h` and do not re-login/redirect users every hour (FedAuth cookie will be valid for `12h` as well)

@Yvand I just want to verify that I correctly undestand this new `sentence` and how AzureCP really works.

**Comment by Yvand** [comment]
The lifetime of the AzureCP access token is totally decorrelated from the user sessions.
If its access token expires in the middle of an active user session (for example, during a POST), AzureCP will merely renew it (which will not cause a sign-out/re-authentication of the user).
User session is valid as long as the FedAuth cookie (and its associated SAML token) is valid, AzureCP never touches this.

**Comment by sergey-tihon** [comment]
Probably I misunderstood something, but I always though that lifetime of `FedAuth` cookie is equal to SAML token lifetime.

As a final step of login SharePoint receive POST request to `_trust/default.aspx` with SAML token
```xml
<t:RequestSecurityTokenResponse xmlns:t="http://schemas.xmlsoap.org/ws/2005/02/trust">
    <t:Lifetime>
        <wsu:Created xmlns:wsu="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd">2020-02-18T16:32:40.915Z</wsu:Created>
        <wsu:Expires xmlns:wsu="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd">2020-02-19T04:37:40.915Z</wsu:Expires>
    </t:Lifetime>
    ...
    <t:TokenType>http://docs.oasis-open.org/wss/oasis-wss-saml-token-profile-1.1#SAMLV1.1</t:TokenType>
    <t:RequestType>http://schemas.xmlsoap.org/ws/2005/02/trust/Issue</t:RequestType>
    <t:KeyType>http://schemas.xmlsoap.org/ws/2005/05/identity/NoProofKey</t:KeyType>
</t:RequestSecurityTokenResponse>
```
SAML is valid for `12h` because I increased my access token lifetime from `1h` to `12h`
In response to this call `default.aspx` set cookie
```
Set-Cookie: FedAuth=77u/...=; expires=Wed, 19-Feb-2020 04:37:43 GMT; path=/; secure; HttpOnly
```
that expires with SAML token.
- FedAuth expires=**Wed, 19-Feb-2020 04:37:43 GMT**
- SAML <wsu:Expires ...>**2020-02-19T04:37:40.915Z**</wsu:Expires>

Is there other way to increase SAML token lifetime (with default value `1h` for access token)?

**Comment by YayForJam** [comment]
I ran into a similar issue today when using the ShareGate migration tool to move a large document library to an offsite SharePoint location that was set up to authenticate using AzureCP. 

Using ShareGate's browser based authentication, it is not able to renew the token automatically so the migration just fails after 50 minutes when the token expires.

I had to create a new Azure AD policy setting the AccessTokenLifetime to 24 hours and assign it to the enterprise application set up for SharePoint and AzureCP. Though the migration it seems is still not going to complete within 24 hours..
We do have a premium Azure AD but as far as I can tell there are no options to simply extend the AccessTokenLifetime for a user or application in the conditional access policies.

**Comment by w3-dev** [comment]
The Sharepoint AzureAD application is using implicit grant with ID tokens for authentication. You might be able to implement the following for the AzureCP

https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-implicit-grant-flow#refreshing-tokens

**Refreshing tokens**
The implicit grant does not provide refresh tokens. Both id_tokens and access_tokens will expire after a short period of time, so your app must be prepared to refresh these tokens periodically. To refresh either type of token, you can perform the same hidden iframe request from above using the prompt=none parameter to control the identity platform's behavior. If you want to receive a new id_token, be sure to use id_token in the response_type and scope=openid, as well as a nonce parameter.

**Comment by Yvand** [comment]
@sergey-tihon initially I thought you were talking about the lifetime of the AzureCP access token, but now I understand you meant the lifetime of the SAML token that users receive from Azure AD upon successful authentication.

@w3-dev AzureCP is not impacted about the 1h lifetime of its access token: When it expires, it just gets a new one and it's completely transparent.

@YayForJam setting AccessTokenLifetime to 24h hours was needed only for the SharePoint enterprise application, it is irrelevant for AzureCP application (see my comment above).

To summarize: the lifetime of the SAML token given by Azure AD to users upon successful authentication is the main (almost only) setting that defines the user session lifetime, and you cannot increase it in SharePoint no matter what configuration you change (you can only lower it). 
AzureCP has no action/impact on the user session lifetime, it uses its own access token that has its own lifetime.

**Comment by Yvand** [comment]
@sergey-tihon any comment/remark? Can this issue be closed?

**Comment by sergey-tihon** [comment]
Yes, you can close it. I'll open new one if we found any issue.
When do you plan to release V16?

**Comment by Yvand** [comment]
I plan to release it very soon, yesterday I published https://github.com/Yvand/AzureCP/releases/tag/16.0.20200303.1092 as a pre-release and it passed all my tests, so it will become the actual release if I don't find any blocking issue.
Are you willing to test it yourself to validate the certificate scenario?

**Comment by sergey-tihon** [comment]
yes, we plan to configure new SP2019 farm soon and I would love to try latest version with cert-based authentication

**Comment by Yvand** [comment]
Perfect, when do you plan to make the test?

**Comment by Yvand** [comment]
@sergey-tihon were you able to test the pre-release?

**Comment by sergey-tihon** [comment]
@Yvand still in-progress. Not working for us, at least for now, but I do not have more details yet

**Comment by sergey-tihon** [comment]
We saw 2 exceptions (during creation of new site collection)
```
[AzureCP] Unexpected error occurred while setting the certificate for tenant 'MYTENANT.onmicrosoft.com'. Is the private key of the certificate exportable?: System.Security.Cryptography.CryptographicException: Keyset does not exist  , Callstack:  
 at System.Security.Cryptography.CryptographicException.ThrowCryptographicException(Int32 hr)   
 at System.Security.Cryptography.X509Certificates.X509Utils._ExportCertificatesToBlob(SafeCertStoreHandle safeCertStoreHandle, X509ContentType contentType, IntPtr password)   
 at System.Security.Cryptography.X509Certificates.X509Certificate.ExportHelper(X509ContentType contentType, Object password)   
 at azurecp.AzureTenant.set_ClientCertificatePrivateKey(X509Certificate2 value)
```

```
[AzureCP] Unexpected error occurred in SearchOrValidate: System.Security.Cryptography.CryptographicException: Keyset does not exist  , Callstack:  
 at System.Security.Cryptography.CryptographicException.ThrowCryptographicException(Int32 hr)   
 at System.Security.Cryptography.X509Certificates.X509Utils._ExportCertificatesToBlob(SafeCertStoreHandle safeCertStoreHandle, X509ContentType contentType, IntPtr password)   
 at System.Security.Cryptography.X509Certificates.X509Certificate.ExportHelper(X509ContentType contentType, Object password)   
 at azurecp.AzureTenant.set_ClientCertificatePrivateKey(X509Certificate2 value)
```

after Googling I found [this answer](https://social.msdn.microsoft.com/Forums/windows/en-US/93916510-7eca-49b3-8260-3337223838c2/cryptographicexception-keyset-does-not-exist?forum=clr)

> **The reason:**
> It is NOT possible to import and use the private key of a PFX certificate from raw data or file within a web service (and probably any IIS web app). I wonder in whose clever head has this idea popped up?! of course nobody cared to mention this exception in the documentation of the appropriate functions. the message "Keyset does not exist" was intended to be self-explanatory ;)
> **The solution:**
> The only way to use your private key with MS Cryptography services within a web service is to have your PFX certificate imported in advance into an X509Store, preferably the local mashine store. Then it becomes easier:

@Yvand how you local environment was configured? What permission IIS/appPool account has on the machine?

**Comment by Yvand** [comment]
I should have provided you with more information to test this, as I had some trouble too.

I tested in 2 ways:

- With a PowerShell script:

```powershell
# Create a new self signed certificate with an exportable private key
$clientCert = New-SelfSignedCertificate -DnsName "AzureCPClientCertificate" -CertStoreLocation "cert:\LocalMachine\My" -KeyExportPolicy Exportable -KeySpec Signature

# Get an existing certificate. To work, its private key must be present and exportable
$clientCert = Get-ChildItem -Path "cert:\localMachine\my" | ?{$_.Subject -like "CN=AzureCPClientCertificate"}

# Update tenant to use the certificate
Add-Type -AssemblyName "AzureCP, Version=1.0.0.0, Culture=neutral, PublicKeyToken=65dc6b5903b51636"
$config = [azurecp.AzureCPConfig]::GetConfiguration("AzureCPConfig")
$tenant = $config.GetAzureTenantByName("TENANTNAME.onMicrosoft.com")
$tenant.UpdateCredentials($clientCert)
$config.Update()
```

- Directly with a .pfx file. I don't remember that I had any problem with a .pfx file though. I assumed it always works since it contains the private key (and I assume is exportable by design).
I created my .pfx files from the certificates MMC console, how did you create yours?

**Comment by sergey-tihon** [comment]
Thanks, we will try this option as well. 

> how did you create yours?

I did it using `openssl`
```bash
openssl genrsa 2048 > azurecp_qa_private.pem
openssl req -x509 -new -key azurecp_qa_private.pem -out azurecp_qa_public.pem -days 365
openssl pkcs12 -export -in azurecp_qa_public.pem -inkey azurecp_qa_private.pem -out azurecp_qa.pfx
```

**Comment by Yvand** [comment]
@sergey-tihon is everything working fine?

**Comment by sergey-tihon** [comment]
@Yvand no, still the same issue.
Can you share how you export certificate from MMC and how you specify cert password for `$tenant`? Did configure private key usage permissions for cert stored in `"cert:\localMachine\my"`?

**Comment by Yvand** [comment]
@sergey-tihon I just tested again the following scenario;

- Create a self-signed certificate in computer store:
`New-SelfSignedCertificate -DnsName "AzureCPClientCertificate" -CertStoreLocation "cert:\LocalMachine\My" -KeyExportPolicy Exportable -KeySpec Signature`
- From the certificates console: I export it as a .pfx with all default options.
- Delete the certificate from the computer store
- Go to AzureCP tenant page > Add a new tenant with the .pfx file

It works.
But I noticed one thing : if I type a wrong password (for the .pfx file), I get exactly the same error as you:

```
[AzureCP] Unexpected error occurred while setting the certificate for tenant 'XXX.onMicrosoft.com'. Is the private key of the certificate exportable?: System.Security.Cryptography.CryptographicException: Keyset does not exist  , Callstack:    at System.Security.Cryptography.CryptographicException.ThrowCryptographicException(Int32 hr)     at System.Security.Cryptography.X509Certificates.X509Utils._ExportCertificatesToBlob(SafeCertStoreHandle safeCertStoreHandle, X509ContentType contentType, IntPtr password)     at System.Security.Cryptography.X509Certificates.X509Certificate.ExportHelper(X509ContentType contentType, Object password)     at azurecp.AzureTenant.set_ClientCertificatePrivateKey(X509Certificate2 value)	c75f3e9f-ff3e-b0a8-ed1e-e4cd036ae1b4
[AzureCP] Unexpected error occurred in SearchOrValidate: System.Security.Cryptography.CryptographicException: Keyset does not exist  , Callstack:    at System.Security.Cryptography.CryptographicException.ThrowCryptographicException(Int32 hr)     at System.Security.Cryptography.X509Certificates.X509Utils._ExportCertificatesToBlob(SafeCertStoreHandle safeCertStoreHandle, X509ContentType contentType, IntPtr password)     at System.Security.Cryptography.X509Certificates.X509Certificate.ExportHelper(X509ContentType contentType, Object password)     at azurecp.AzureTenant.set_ClientCertificatePrivateKey(X509Certificate2 value)	c75f3e9f-ff3e-b0a8-ed1e-e4cd036ae1b4
```

Could it be the same issue in your case?

**Comment by sergey-tihon** [comment]
Hmm, where do you persist `ClientCertificatePrivateKeyPassword` ? I still see hardcoded value in `dev` branch, but cannot find line where you persist `password` that we type when upload certificate.

**Comment by Yvand** [comment]
ClientCertificatePrivateKeyPassword is not persisted, it's a const used to export/import the certificate (with its private key) that is persisted in the configuration database.
The password you type in the configuration page is used only to decrypt the .pfx file (it assumes it is password protected).
It looks like your .pfx file is not password protected, and importing it fails because it expects a password is required.

**Comment by Yvand** [comment]
@sergey-tihon can you please update/comment? I'd like to publish the v16, unless you find a blocking issue

**Comment by sergey-tihon** [comment]
@Yvand Sorry, for my slow response here.
 I've tried to find `repro` steps for, but no success for now.

It works for some period of time (after tenant creation), so I am able to create site collection choosing site collection admin from picker, I see successful communication with Graph. 
```
[AzureCP] Configuration 'AzureCPConfig' changed to version 442001, refreshing local copy	[AzureCP] Refreshing local copy of configuration 'AzureCPConfig'	
[AzureCP] Getting new access token for tenant 'XXX.onmicrosoft.com' using client ID ___ and a client certificate with thumbprint B3273ACB4928A3F580A3DE755E3F477C7DD6F6AA.		[AzureCP] Got new access token for tenant 'XXX.onmicrosoft.com', valid for 0.9 hour(s) and retrieved in 0 ms
[AzureCP] Got 1 users/groups in 297 ms from 'XXX.onmicrosoft.com' with input 'sergey_tihon'	
[AzureCP] 1 entity(ies) to create after filtering	db5b409f-02dc-70fb-c3d5-0382883f195a
```

after some period of time (here is inconsistent) when I try to change site collection admin to another person I see
```
[AzureCP] Configuration 'AzureCPConfig' changed to version 442001, refreshing local copy	[AzureCP] Refreshing local copy of configuration 'AzureCPConfig'	
[AzureCP] Unexpected error occurred while setting the certificate for tenant 'XXX.onmicrosoft.com'. Is the private key of the certificate exportable?: System.Security.Cryptography.CryptographicException: Keyset does not exist  , Callstack:    at System.Security.Cryptography.CryptographicException.ThrowCryptographicException(Int32 hr)     at System.Security.Cryptography.X509Certificates.X509Utils._ExportCertificatesToBlob(SafeCertStoreHandle safeCertStoreHandle, X509ContentType contentType, IntPtr password)     at System.Security.Cryptography.X509Certificates.X509Certificate.ExportHelper(X509ContentType contentType, Object password)     at azurecp.AzureTenant.set_ClientCertificatePrivateKey(X509Certificate2 value)
[AzureCP] Unexpected error occurred in SearchOrValidate: System.Security.Cryptography.CryptographicException: Keyset does not exist  , Callstack:    at System.Security.Cryptography.CryptographicException.ThrowCryptographicException(Int32 hr)     at System.Security.Cryptography.X509Certificates.X509Utils._ExportCertificatesToBlob(SafeCertStoreHandle safeCertStoreHandle, X509ContentType contentType, IntPtr password)     at System.Security.Cryptography.X509Certificates.X509Certificate.ExportHelper(X509ContentType contentType, Object password)     at azurecp.AzureTenant.set_ClientCertificatePrivateKey(X509Certificate2 value)
```
I have no idea why ...


> it's a const used to export/import the certificate (with its private key) that is persisted in the configuration database.

Can `export/import` be process specific? Any issues that you can imaging with importing private key from DB?

**Comment by Yvand** [comment]
@sergey-tihon sorry I completely forgot to reply.
I didn't repro this issue but I probably didn't try as much as you. 
I don't have a clue about why it occurs, the only thing I find strange is that it tries to re-import the certificate because it of a config change, but the version is the same so it didn't actually change.

Could it happen somehow when the access token expires? 
I need to try this in my environment

**Comment by Yvand** [comment]
The renewal of access token works as expected in my environment. When it happens, the following is recorded in the logs:

```
AzureCP	Core	1337	Verbose	[AzureCP] Access token for tenant 'XXX.onMicrosoft.com' expired, renewing it	167c449f-bb52-a0a1-49c7-0aea33aee95f
AzureCP	Core	1337	Verbose	[AzureCP] Getting new access token for tenant 'XXX.onMicrosoft.com' using client ID <GUID> and a client certificate with thumbprint A03D77775D7C90F4703473356ABBD7B0E418BDAA.	167c449f-bb52-a0a1-49c7-0aea33aee95f
AzureCP	Core	1337	High	[AzureCP] Got new access token for tenant 'XXX.onMicrosoft.com', valid for 1 hour(s) and retrieved in 1784 ms	
```

You do not seem to be in that scenario. For some reason, the AzureCP config is refreshed with no reason (since it's the same version), that should not happen on a process that has already cached that config version.

**Comment by sergey-tihon** [comment]
how can I help you troubleshoot this? IIRC, I saw AzureCP messages in ULS logs from multiple processes, not only w3wp but able from owstimer. Can User Profile Server somehow use AzureCP?

**Comment by Yvand** [comment]
My understanding of the issue is the following:
In a given process instance, AzureCP works fine for some time until, for some reason, it refreshes its config despite the version didn't change, and fails at the step of importing the certificate.
Is this correct? Is it always this behavior?

If so, can you provide the AzureCP logs which illustrate exactly this? They should contain the following sequence:

- A refresh of the config for a given version (ex. v1)
- Access token using certificate successfully retrieved
- A successful Graph operation (a dump search in the people picker)
- The unexpected refresh of the config with the same version (ex. v1)
- And the error when importing the private key

If you don't mind, when you have this data, can you open a new issue? This one has become too big and was for a diffent topic.

About the w3wp of the User Profile Service application: I think it may load the assembly, but I don't think it will actually use it, although I'm not sure.

**Comment by sergey-tihon** [comment]
Ok, I will try
